lightning-transaction-sync/src/esplora.rs

```diff
@@ -0,0 +1,424 @@
+		// reconfirmed in another block, we'll confirm them in the next sync iteration.
+		let relevant_txids = confirmables
+			.iter()
+			.flat_map(|c| c.get_relevant_txids())
+			.collect::<HashSet<(Txid, Option<BlockHash>)>>();
+
+		for (txid, block_hash_opt) in relevant_txids {
+			if let Some(block_hash) = block_hash_opt {
+				let block_status = maybe_await!(self.client.get_block_status(&block_hash))?;
+				if block_status.in_best_chain {
+					// Skip if the block in question is still confirmed.
+					continue;
+				}
+			}
+
+			for c in confirmables {
+				c.transaction_unconfirmed(&txid);
+			}
+
+			sync_state.watched_transactions.insert(txid);
+		}
+
+		Ok(())
+	}
+
+	/// Returns a reference to the underlying esplora client.
+	pub fn client(&self) -> &EsploraClientType {
+		&self.client
+	}
+}
+
+// Represents the current state.
+struct SyncState {
+	// Transactions that were previously processed, but must not be forgotten
+	// yet since they still need to be monitored for confirmation on-chain.
+	watched_transactions: HashSet<Txid>,
+	// Outputs that were previously processed, but must not be forgotten yet as
+	// as we still need to monitor any spends on-chain.
+	watched_outputs: HashSet<WatchedOutput>,
+	// The tip hash observed during our last sync.
+	last_sync_hash: Option<BlockHash>,
+	// Indicates whether we need to resync, e.g., after encountering an error.
+	pending_sync: bool,
+}
+
+impl SyncState {
+	fn new() -> Self {
+		Self {
+			watched_transactions: HashSet::new(),
+			watched_outputs: HashSet::new(),
+			last_sync_hash: None,
+			pending_sync: false,
+		}
+	}
+}
+
+// A queue that is to be filled by `Filter` and drained during the next syncing round.
+struct FilterQueue {
+	// Transactions that were registered via the `Filter` interface and have to be processed.
+	transactions: HashSet<Txid>,
+	// Outputs that were registered via the `Filter` interface and have to be processed.
+	outputs: HashSet<WatchedOutput>,
+}
+
+impl FilterQueue {
+	fn new() -> Self {
+		Self {
+			transactions: HashSet::new(),
+			outputs: HashSet::new(),
+		}
+	}
+
+	// Processes the transaction and output queues and adds them to the given [`SyncState`].
+	//
+	// Returns `true` if new items had been registered.
+	fn process_queues(&mut self, sync_state: &mut SyncState) -> bool {
+		let mut pending_registrations = false;
+
+		if !self.transactions.is_empty() {
+			pending_registrations = true;
+
+			sync_state.watched_transactions.extend(self.transactions.drain(..));
+		}
+
+		if !self.outputs.is_empty() {
+			pending_registrations = true;
+
+			sync_state.watched_outputs.extend(self.outputs.drain(..));
+		}
+		pending_registrations
+	}
+}
+
+#[cfg(feature = "async-interface")]
+type MutexType<I> = futures::lock::Mutex<I>;
+#[cfg(not(feature = "async-interface"))]
+type MutexType<I> = std::sync::Mutex<I>;
+
+/// The underlying client type.
+#[cfg(feature = "async-interface")]
+pub type EsploraClientType = AsyncClient;
```

Comment:
While I really like that we now can switch the entire interface between blocking / async based on a simple feature, it currently means that the docs only show one of the variants. E.g., `cargo doc --all-features` produces the async variant for me. 

@jkczyz Any thoughts? Are we fine with that as long as we sufficiently document the feature use in the crate docs?

Reply:
Hmm... not sure I follow how the newtype variation would work. Could you expand upon that?

Also, how does `#[maybe_async]` affect the docs?

Reply:
Well a newtype for `EsploraClientType` would allow us to even document it, currently the type alias is simply replaced in the docs. Likewise, `#[maybe_async]` just switches out the impl based on the feature flag.

That means that the entire interface switches in-place, resulting in the only one of the versions being documented (`async` for me, presumably the last one tried overrides the previously generated versions). 

The alternative would be to copy-paste the entire module and create two separate `EsploraAsyncSyncClient`/`EsploraBlockingSyncClient` modules, which I'm however not a big fan of.

Reply:
> Well a newtype for `EsploraClientType` would allow us to at least document it, currently the type alias is simply replaced/resolved in the docs. Likewise, `#[maybe_async]` just switches out the impl based on the feature flag.
> 

I mean, I don't follow how this is much different from a type alias from a documentation prespective. Are you saying something different from the following?

```rust
#[cfg(feature = "async-interface")]
pub struct EsploraClientType(AsyncClient);
#[cfg(not(feature = "async-interface"))]
pub struct EsploraClientType(BlockingClient);
```

> That means that the entire interface switches in-place, resulting in the only one of the versions being documented (`async` for me, presumably the last one tried overrides the previously generated versions).
> 

I think it's the `async` version because `cargo doc --all-features` will *always* set `async-interface` whereas the blocking version is only possible when `async-interface` is *not* set (i.e., there isn't a separate feature for blocking).

> The alternative would be to copy-paste the entire module and create two separate `EsploraAsyncSyncClient`/`EsploraBlockingSyncClient` modules, which I'm however not a big fan of.

Yeah, I don't see any other alternative and wouldn't find that desirable either... Though is there any way to have some internal module using parameterized types that are exposed as two separate modules using different type parameterizations? I'm thinking more like `lightning-transaction-sync::blocking::EsploraSyncClient` and `lightning-transaction-sync::async::EsploraSyncClient` where those are type aliases for the corresponding parameterizations. Guessing the use of `[#maybe_async]` would make that challenging / impossible?

Reply:
I haven't seen a good solution to this. Not at my computer right now, but `esplora-client` seems to take the approach of copying the entire module and exposing async and blocking variants, while bdk just defaults to their blocking API (which admittedly can be quite confusing for eg a reviewer ("how'd you make this async?"))

Maybe the complex common logic could be extracted into a private module, and the blocking and async modules are just interfaces into it, reducing the amount of duplication to just the interfaces?

Reply:
> I mean, I don't follow how this is much different from a type alias from a documentation prespective. Are you saying something different from the following?
> 
> ```rust
> #[cfg(feature = "async-interface")]
> pub struct EsploraClientType(AsyncClient);
> #[cfg(not(feature = "async-interface"))]
> pub struct EsploraClientType(BlockingClient);
> ```
No, I meant exactly that, which would make `EsploraClientType` itself show up in the docs. Currently, the type alias is directly resolved and the methods show up as:
<img width="602" alt="Screenshot 2023-01-24 at 09 51 34" src="https://user-images.githubusercontent.com/521467/214341749-02e1d8e2-0d50-4fdc-9905-32b3ddc533b8.png">

(Which I have to say I don't find too bad, probably preferable over introducing a newtype that would only barely improve it but would be an additional stepping stone for the user)

> I think it's the `async` version because `cargo doc --all-features` will _always_ set `async-interface` whereas the blocking version is only possible when `async-interface` is _not_ set (i.e., there isn't a separate feature for blocking).

But there is, we have two features, `esplora-blocking` and `esplora-async`, which both have to be deliberately activated by the user.

> Yeah, I don't see any other alternative and wouldn't find that desirable either... Though is there any way to have some internal module using parameterized types that are exposed as two separate modules using different type parameterizations? I'm thinking more like `lightning-transaction-sync::blocking::EsploraSyncClient` and `lightning-transaction-sync::async::EsploraSyncClient` where those are type aliases for the corresponding parameterizations. Guessing the use of `[#maybe_async]` would make that challenging / impossible?

I suppose we could work around that with something along the lines of what Max suggested. However, that really sounds like a lot of added complexity and boilerplate for a doc issue that may be also be resolved by sufficient `Note`s in the crate and module docs?

Reply:
> I suppose we could work around that with something along the lines of what Max suggested. However, that really sounds like a lot of added complexity and boilerplate for a doc issue that may be also be resolved by sufficient `Note`s in the crate and module docs?

Could use a macro... 😉 

In all seriousness, the public API is fairly small, it seems. So Max's suggestion may not be horrible. That said, I don't have strong opinion on this. In fact, `lightning-block-sync` only exposes an `async` interface as it is. Though users of `lightning-transaction-sync` may want the blocking option.

Reply:
SGTM